backends/bmv2/common/expression.cpp

```diff
@@ -184,32 +184,51 @@ void ExpressionConverter::postorder(const IR::Constant* expression)  {
 
 void ExpressionConverter::postorder(const IR::ArrayIndex* expression)  {
     auto result = new Util::JsonObject();
-    result->emplace("type", "header");
     cstring elementAccess;
 
+    LOG2("left: " << expression->left << " right: " << expression->right);
     // This is can be either a header, which is part of the "headers" parameter
     // or a temporary array.
-    if (expression->left->is<IR::Member>()) {
+    if (auto mem = expression->left->to<IR::Member>()) {
         // This is a header part of the parameters
-        auto mem = expression->left->to<IR::Member>();
         auto parentType = typeMap->getType(mem->expr, true);
         BUG_CHECK(parentType->is<IR::Type_StructLike>(), "%1%: expected a struct", parentType);
         auto st = parentType->to<IR::Type_StructLike>();
         auto field = st->getField(mem->member);
         elementAccess = field->controlPlaneName();
-    } else if (expression->left->is<IR::PathExpression>()) {
+    } else if (auto path = expression->left->to<IR::PathExpression>()) {
         // This is a temporary variable with type stack.
-        auto path = expression->left->to<IR::PathExpression>();
         elementAccess = path->path->name;
     }
 
     if (!expression->right->is<IR::Constant>()) {
-        ::error(ErrorType::ERR_INVALID, "all array indexes must be constant", expression->right);
+        const IR::Expression* ex = expression->right;
+        auto res = map.find(ex);
+        Util::JsonObject* fres;
+        if (res != map.end()) {
```

Comment:
Please use the pattern used everywhere else: `get(expression)`, it does it's own error reporting.